### ISSUE TITLE
OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.(Using Tensorflow data pipeline map function with tfrecords)

### ISSUE THREAD

**Issue by bhupendrathore** [tensorflow/tensorflow#44538]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1.  It must be a bug, a feature request, or a significant problem with the
    documentation (for small docs fixes please send a PR instead).
2.  The form below must be filled out.
3.  It shouldn't be a TensorBoard issue. Those go
    [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04.5
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**: NO
-   **TensorFlow installed from (source or binary)**: pip install tf-nightly 
-   **TensorFlow version (use command below)**:'2.4.0-dev20201019'
-   **Python version**: 3.7.5
-   **Bazel version (if compiling from source)**:
-   **GCC/Compiler version (if compiling from source)**:
-   **CUDA/cuDNN version**: NA
-   **GPU model and memory**:NA
-   **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
I converted ms-coco format data into tfrecords, loaded the same using   

 autotune = tf.data.experimental.AUTOTUNE
data = tf.data.TFRecordDataset(path)
data = data.map(_decode_record,num_parallel_calls=autotune)
data = data.map(preprocess_data,num_parallel_calls= autotune)
## I've pasted the functions used below in the source code block

preprocess function have some condition check, but when using map all are executed in eager mode so gives error :

OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.

tried @tf.function, tf.py_function, functool.partial but no success, Before using tf nightly it was throwing error : Tensor' object has no attribute 'numpy'

### Source code / logs
###Creating tfrecords code:
with tf.io.TFRecordWriter(my_path) as writer:
    def _bytes_feature(value):
        if isinstance(value, type(tf.constant(0))):
            value = value.numpy()
        return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))
    def _convert_and_serialize(value):
        value = tf.convert_to_tensor(value)
        value = tf.io.serialize_tensor(value)
        return value
    image = np.asarray(PIL.Image.open(img_path))
   feature = {}
    image = _convert_and_serialize(image)
   feature['image'] = _bytes_feature(image)tf_example = tf.train.Example(features=tf.train.Features(feature=feature))
   writer.write(tf_example.SerializeToString())

### Reading tfrecords 
"""
def _decode_record(record):
    feature_description = {
    'image': tf.io.FixedLenFeature([], tf.float64, default_value=0),
    }
    parsed = tf.io.parse_single_example(record,feature_description)
    return parsed
"""
#preprocess function
"""
def random_flip_horizontal(image, boxes):
    """Flips image and boxes horizontally with 50% chance

    Arguments:
      image: A 3-D tensor of shape `(height, width, channels)` representing an
        image.
      boxes: A tensor with shape `(num_boxes, 4)` representing bounding boxes,
        having normalized coordinates.

    Returns:
      Randomly flipped image and boxes
    """
    if tf.random.uniform(()) > 0.5:
        image = tf.image.flip_left_right(image)
        boxes = tf.stack(
            [1 - boxes[:, 2], boxes[:, 1], 1 - boxes[:, 0], boxes[:, 3]], axis=-1
        )
    return image, boxes

def resize_and_pad_image(
    image, min_side=800.0, max_side=1333.0, jitter=[640, 1024], stride=128.0
):
    """Resizes and pads image while preserving aspect ratio.

    1. Resizes images so that the shorter side is equal to `min_side`
    2. If the longer side is greater than `max_side`, then resize the image
      with longer side equal to `max_side`
    3. Pad with zeros on right and bottom to make the image shape divisible by
    `stride`

    Arguments:
      image: A 3-D tensor of shape `(height, width, channels)` representing an
        image.
      min_side: The shorter side of the image is resized to this value, if
        `jitter` is set to None.
      max_side: If the longer side of the image exceeds this value after
        resizing, the image is resized such that the longer side now equals to
        this value.
      jitter: A list of floats containing minimum and maximum size for scale
        jittering. If available, the shorter side of the image will be
        resized to a random value in this range.
      stride: The stride of the smallest feature map in the feature pyramid.
        Can be calculated using `image_size / feature_map_size`.

    Returns:
      image: Resized and padded image.
      image_shape: Shape of the image before padding.
      ratio: The scaling factor used to resize the image
    """
    image_shape = tf.cast(tf.shape(image)[:2], dtype=tf.float32)
    if jitter is not None:
        min_side = tf.random.uniform((), jitter[0], jitter[1], dtype=tf.float32)
    ratio = min_side / tf.reduce_min(image_shape)
    if ratio * tf.reduce_max(image_shape) > max_side:
        ratio = max_side / tf.reduce_max(image_shape)
    image_shape = ratio * image_shape
    image = tf.image.resize(image, tf.cast(image_shape, dtype=tf.int32))
    padded_image_shape = tf.cast(
        tf.math.ceil(image_shape / stride) * stride, dtype=tf.int32
    )
    image = tf.image.pad_to_bounding_box(
        image, 0, 0, padded_image_shape[0], padded_image_shape[1]
    )
    return image, image_shape, ratio

def preprocess_data(sample):
    """Applies preprocessing step to a single sample

    Arguments:
      sample: A dict representing a single training sample.

    Returns:
      image: Resized and padded image with random horizontal flipping applied.
      bbox: Bounding boxes with the shape `(num_objects, 4)` where each box is
        of the format `[x, y, width, height]`.
      class_id: An tensor representing the class id of the objects, having
        shape `(num_objects,)`.
    """
    image = sample["image"]
    bbox = swap_xy(sample["objects"]["bbox"])
    class_id = tf.cast(sample["objects"]["label"], dtype=tf.int32)

    image, bbox = random_flip_horizontal(image, bbox)
    image, image_shape, _ = resize_and_pad_image(image)

    bbox = tf.stack(
        [
            bbox[:, 0] * image_shape[1],
            bbox[:, 1] * image_shape[0],
            bbox[:, 2] * image_shape[1],
            bbox[:, 3] * image_shape[0],
        ],
        axis=-1,
    )
    bbox = convert_to_xywh(bbox)
    return image, bbox, class_id
"""

#error log 

<ipython-input-66-b4ebcba6c559> in preprocess_data(sample)
     91 
     92 #     image, bbox = random_flip_horizontal(image, bbox)
---> 93     image, image_shape, _ = resize_and_pad_image(image)
     94     image_shape = image.shape
     95 

<ipython-input-66-b4ebcba6c559> in resize_and_pad_image(image, min_side, max_side, jitter, stride)
     58     print('ratio',ratio,type(ratio))
     59     print(tf.reduce_max(image_shape),type(tf.reduce_max(image_shape)))
---> 60     if ratio * tf.reduce_max(image_shape) > max_side:
     61         ratio = max_side / tf.reduce_max(image_shape)
     62     image_shape = ratio * image_shape

env/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in __bool__(self)
    883       `TypeError`.
    884     """
--> 885     self._disallow_bool_casting()
    886 
    887   def __nonzero__(self):

env/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in _disallow_bool_casting(self)
    490     else:
    491       # Default: V1-style Graph execution.
--> 492       self._disallow_in_graph_mode("using a `tf.Tensor` as a Python `bool`")
    493 
    494   def _disallow_iteration(self):

env/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in _disallow_in_graph_mode(self, task)
    479     raise errors.OperatorNotAllowedInGraphError(
    480         "{} is not allowed in Graph execution. Use Eager execution or decorate"
--> 481         " this function with @tf.function.".format(task))
    482 
    483   def _disallow_bool_casting(self):

OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.

If there is better way to do the same then also let me know.

Basically i wanted to replicate the same as given below example : https://keras.io/examples/vision/retinanet/
But with tf-nightly this also gives same error.

Thanks,

**Comment by Saduf2019** [comment]
@bhupendrathore 
code shared is not indeneted, please provide a colab gist with the error reported.also can you verify if the error exist in nightly.

**Comment by bhupendrathore** [comment]
Hi @Saduf2019,

Tried to reproduce the same error but could not install tf-nightly same version (2.4.0-dev20201019),

gist can be found at this URL : https://gist.github.com/bhupendrathore/c7c737e53a3917a87c27b443d3620ee5

colab also can be found at this URL : https://colab.research.google.com/drive/1-c4OYKe42pMqouVJAFcznhIfv8GKOVSH?usp=sharing

**Comment by Saduf2019** [comment]
@bhupendrathore 
I ran the code shared there are too many errors, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/a42035b1419048093449eac1fc8381c9/untitled458.ipynb)

**Comment by bhupendrathore** [comment]
@Saduf2019 find the updated [colab notebook here](https://colab.research.google.com/drive/1-c4OYKe42pMqouVJAFcznhIfv8GKOVSH#scrollTo=Q8kksBJoQ5Bs)

now the error has changed with new tf-nightly which is 
'''
ValueError: in user code:

    <ipython-input-52-b4ebcba6c559>:93 preprocess_data  *
        image, image_shape, _ = resize_and_pad_image(image)
    <ipython-input-58-b4ebcba6c559>:63 resize_and_pad_image  *
        image = tf.image.resize(image, tf.cast(image_shape, dtype=tf.int32))
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/util/dispatch.py:206 wrapper  **
        return target(*args, **kwargs)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/image_ops_impl.py:1647 resize_images_v2
        skip_resize_if_same=False)
    /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/image_ops_impl.py:1313 _resize_images_common
        raise ValueError('\'images\' contains no shape.')

    ValueError: 'images' contains no shape.
'''
It's same as error with tf 2.3, to avoid i used tf-nightly so in the version (2.4.0-dev20201019) i was getting the tf.Tensor as a python bool is not allowed in graph execution but now with the latest tf-nightly its same as images contains no shape.

Suggest if you want me to close this and reopen new, but I believe both are because of the execution in eager mode and image tensor not able to find shape attribute in eager mode.
correct me if I'm wrong
Thanks

**Comment by Saduf2019** [comment]
I am able to replicate the issue faced, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/dee1694852cf01f77a51df342a1964d0/untitled458.ipynb)

**Comment by jvishnuvardhan** [comment]
@bhupendrathore Can you please provide simple standalone code to reproduce the issue. The current code is too long to support. Thanks!

**Comment by jvishnuvardhan** [comment]
Sorry, by mistake I closed. Reopened. Thanks!
